:memo: Rework activity reference (#2766)
platformsh/platformsh-docs: docs/src/integrations/activity/reference.md

  "completed_at": "2022-12-16T14:31:48.809068+00:00",
  "completion_percent": 100,
  "cancelled_at": null,
  "timings": {
    "wait": 0,
    "build": 0.349,
    "deploy": 209.986,
    "execute": 210.508
  },
  "log": "Building application 'app' (runtime type: php:8.2, tree: 9851a01)\n  Reusing existing build for this tree ID\n\n...\nRedeploying environment test, as a clone of main\n  ...\n  Closing all services\n  Opening application app and its relationships\n  Executing deploy hook for application app\n ... Environment configuration\n    app (type: php:8.0, size: S, disk: 2048)\n\n  ...",
  "payload": {
    ...
  },
  "description": "<user data-id=\"abcdefghijk123456789\">Cloé Weber</user> synchronized <environment data-id=\"test\">test</environment>'s <strong>data</strong> from <environment data-id=\"main\">Main</environment>",
  "text": "Cloé Weber synchronized test's **data** from Main",
  "expires_at": "2023-12-16T14:28:17.890467+00:00"
}
```

### `id`

A unique `id` value to identify the activity itself.

### `*_at`

`created_at`, `started_at`, `updated_at`, `cancelled_at`, `completed_at`, and `expires_at` are all timestamps in UTC.

> For when a given action happened, use `completed_at`.

Reviewer: ```suggestion
For when a given activity occurred, use `completed_at`.
```

This makes it unlikely that activities on non-production environments block activities for the production environment,
though there may be a temporary wait.

## Examples

The response is often usually long, so the following examples are shortened using ellipses.
Remember that the response differs depending on the activity and not all fields are always available.

To test responses, [set up a webhook](./webhooks.md#setup).

### Cron

When a cron job is triggered, the activity contains all the [job's information](../../create-apps/app-reference.md#crons).
In that example, the cron is scheduled to run every five minutes (`5 * * * *`),
runs `sleep 60 && echo sleep-60-finished && date` and timeouts after 86400 seconds.

To get details about your cron, see the `parameters` property:

``` json
...
  "parameters": {
    "user": "admin",
    "cluster": "abcdefgh1234567-main-abcd123",
    "environment": "main",
    "application": "app",
    "cron": "saybye",
    "spec": {
      "spec": "5 * * * *",
      "commands": {
        "start": "sleep 60 && echo sleep-60-finished && date",
        "stop": null
      },
      "shutdown_timeout": null,
      "timeout": 86400
    }
...
```

> The following presents the full activity response to a cron:

Reviewer: ```suggestion
The following example shows the full activity response to a cron job:
```

      },
      "resources_overrides": {},
      "last_active_at": "2022-12-13T15:07:10.862854+00:00",
      "last_backup_at": null,
      "project": "abcdefgh1234567",
      "is_main": true,
      "is_dirty": false,
      "has_code": true,
      "head_commit": "6aac318907b50252976c47e4e62ed95d438af0ea",
      "merge_info": {
        "commits_ahead": 0,
        "commits_behind": 0,
        "parent_ref": null
      },
      "has_deployment": true
    },
    "cron": "saybye"
  },
  "description": "<user data-id=\"admin\">Platform.sh Bot</user> ran cron <strong>saybye</strong>",
  "text": "Platform.sh Bot ran cron **saybye**",
  "expires_at": "2023-01-12T16:06:08.081293+00:00"
}
```

### Push

A push activity contains several properties.
The `commits` property contains everything related to the git push you made:

``` json
...
    "commits": [
      {
        "sha": "2bab04e050279ac078d5d34016f5dd9c466e948d",
        "author": {
          "email": "cloeweber@example.com",
          "name": "Cloé Weber",
          "date": 1671032461
        },
        "parents": [
          "6aac318907b50252976c47e4e62ed95d438af0ea"
        ],
        "message": "Add cron"
      }
    ],
...
```

> The `environment` property contains your environment's settings:

Reviewer: ```suggestion
The `environment` property contains the settings for the environment that was pushed to:
```

      "updated_at": "2022-11-29T16:16:37.085719+00:00",
      "name": "main",
      "machine_name": "main-abcd123",
      "title": "Main",
      "attributes": {},
      "type": "production",
      "parent": null,
      "default_domain": null,
      "clone_parent_on_create": true,
      "deployment_target": "local",
      "is_pr": false,
      "status": "active",
      "enable_smtp": true,
      "restrict_robots": true,
      "edge_hostname": "main-abcd123-abcdefgh1234567.eu-3.platformsh.site",
      "deployment_state": {
        "last_deployment_successful": true,
        "last_deployment_at": "2022-11-29T16:16:37.085609+00:00",
        "crons": {
          "enabled": true,
          "status": "sleeping"
        }
      },
      "resources_overrides": {},
      "last_active_at": "2022-12-13T16:07:09.788910+00:00",
      "last_backup_at": null,
      "project": "abcdefgh1234567",
      "is_main": true,
      "is_dirty": false,
      "has_code": true,
      "head_commit": "6aac318907b50252976c47e4e62ed95d438af0ea",
      "merge_info": {
        "commits_ahead": 0,
        "commits_behind": 0,
        "parent_ref": null
      },
      "has_deployment": true
    },
...
```

> The `deployment` property contains the settings used such as the [image type](../../create-apps/app-reference.md#types),
> the [resource allocation](../../create-apps/app-reference.md#sizes) and much more.
> 
> A shortened excerpt of the `deployment` property looks like:

Reviewer: ```suggestion
The `deployment` property contains the settings for the deployment,
including the [image type](../../create-apps/app-reference.md#types) and
[resource allocation](../../create-apps/app-reference.md#sizes).

The following example shows a shortened excerpt of the `deployment` property:
```

                "files": [],
                "profile": "PHP",
                "config": null,
                "repository": "https://github.com/platformsh-templates/php.git@master",
                "title": "PHP"
              }
            }
          },
          ...
          "application_config_file": ".platform.app.yaml",
          "project_config_dir": ".platform",
          ...
          "development_service_size": "S",
          "development_application_size": "S",
          "enable_certificate_provisioning": true,
          "certificate_style": "ecdsa",
          "certificate_renewal_activity": true,
          ...
          "cron_minimum_interval": 5,
          "cron_maximum_jitter": 20,
          "concurrency_limits": {
            "internal": null,
            "integration": 4,
            "backup": 2,
            "cron": 10,
            "default": 2
          },
          ...
          "build_resources": {
            "cpu": 1,
            "memory": 2048
          },
          ...
          "max_allowed_routes": 50000,
          "max_allowed_redirects_paths": 50000,
          "enable_incremental_backups": true,
          ...
        }
      },
...
```

> The following presents the full activity response to a Git push:

Reviewer: ```suggestion
The following example shows the full activity response to a Git push:
```

| `payload.project` | Information about the project. For details on its properties, see the [`project` payload](#project-payload). |

#### `user` payload

Contains information about the Platform.sh user that triggered the activity.

| Name | Description |
|------|-------------|
| `payload.user.created_at` | The date the user was created. |
| `payload.user.display_name` | The user's name in a human-friendly format. |
| `payload.user.id` | The user's ID. |
| `payload.user.updated_at` | The date the user was last updated. |

#### `environment` payload

Contains information about the environment associated with the activity,
including its settings, state, and deployment.
The following table presents the most notable properties of the environment:

| Name | Description |
|------|-------------|
| `payload.environment.name` | The environment name. |
| `payload.environment.type` | The [environment type](../../administration/users.md#environment-types). |
| `payload.environment.head_commit` | The ID of the environment's latest Git commit. |

> | `payload.environment.edge_hostname` | The URL you should target when setting up a [custom domain](../../domains/steps.md). |

Reviewer: ```suggestion
| `payload.environment.edge_hostname` | The URL you should target when setting up a [custom domain](../../domains/steps/_index.md). |
```

title: "Activity reference"
sidebarTitle: "Activity reference"
weight: -10
description: Reference for the activities
---

> Activities are responses to actions in your project,

Author: Should be more ocnsistent now :)

including when you deploy your app,
when you [push code](#push) or when a [cron job is run](#cron).

## Activity schema

Any activity has a JSON object linked to it containing all information for that activity such as timestamps,

> duration, configuration, and much more.

Author: What do you mean exactly with "Maybe logs worth highlighting?"?

> duration, configuration, and much more.

Reviewer: I meant, maybe it'd be worth highlighting the availability of logs in the response. To replace the duration :slightly_smiling_face:

The JSON object can be parsed and reacted to through custom hooks.
You can use these hooks to automate your workflows.

Depending on the activity, the JSON object response differs and not all fields are always available.

In [JSON terminology](https://json-schema.org/draft-04/json-schema-core.html#rfc.section.3),
"property" refers to a JSON object and "element" refers to a JSON array.

See full examples of [the activity schema](#examples).

### Example response

Here is a shortened example of what a JSON object response looks like for the [sync of an environment](../../other/glossary.md#sync):

``` json
{
  "id": "abcdefg123456",
  ...
  "created_at": "2022-12-16T14:28:17.890467+00:00",
  "updated_at": null,
  "type": "environment.synchronize",
  "parameters": {
    ...
  },
  "project": "abcdefgh1234567",
  "environments": [
    "feature"
  ],
  "state": "complete",
  "result": "success",
  "started_at": "2022-12-16T14:28:18.188888+00:00",
  "completed_at": "2022-12-16T14:31:48.809068+00:00",
  "completion_percent": 100,
  "cancelled_at": null,
  "timings": {
    "wait": 0,
    "build": 0.349,
    "deploy": 209.986,
    "execute": 210.508
  },

>   "log": "Building application 'app' (runtime type: php:8.2, tree: 9851a01)\n  Reusing existing build for this tree ID\n\nProvisioning certificates\n  Certificates\n  - certificate 5684968: expiring on 2023-03-15 13:25:28+00:00, covering {,www}.feature-abcd123-abcdefgh1234567.eu.platformsh.site\n\n\nRedeploying environment test, as a clone of main\n  Taking a temporary backup of the parent environment\n  Preparing to restore from a temporary backup\n  Preparing deployment\n  Closing all services\n  Opening application app and its relationships\n  Executing deploy hook for application app\n    hello world\n\n  Opening environment\n  Deleting temporary backups (1)\n  Environment configuration\n    app (type: php:8.0, size: S, disk: 2048)\n\n  Environment routes\n    http://feature-abcd123-abcdefgh1234567.eu.platformsh.site/ redirects to https://feature-abcd123-abcdefgh1234567.eu.platformsh.site/\n    http://www.feature-abcd123-abcdefgh1234567.eu.platformsh.site/ redirects to https://www.feature-abcd123-abcdefgh1234567.eu.platformsh.site/\n    https://feature-abcd123-abcdefgh1234567.eu.platformsh.site/ is served by application `app`\n    https://www.feature-abcd123-abcdefgh1234567.eu.platformsh.site/ redirects to https://feature-abcd123-abcdefgh1234567.eu.platformsh.site/\n",

Reviewer: :shrug: It's probably OK.

  "payload": {
    ...
  },
  "description": "<user data-id=\"abcdefghijk123456789\">Cloé Weber</user> synchronized <environment data-id=\"test\">test</environment>'s <strong>data</strong> from <environment data-id=\"main\">Main</environment>",
  "text": "Cloé Weber synchronized test's **data** from Main",
  "expires_at": "2023-12-16T14:28:17.890467+00:00"
}
```

### `id`

A unique `id` value to identify the activity itself.

### `*_at`

The `created_at`, `started_at`, `updated_at`, `cancelled_at`, `completed_at`, `expires_at` values are all timestamps in UTC.
If you need only a point in time when a given action happened, use `completed_at`.
If needed, you can use these properties to compute the duration of the different steps of a task.
For more time computation, [see `timings`](#timings).

### `parameters`

The parameters property include detailed information regarding which user ID triggered the activity,
the environment impacted, the ID of the git commits, commands run with crons, or if changes in the codebase happened.

> The response changes based on the activity.

Reviewer: Maybe out of scope?

### `project`

The ID value of the project in which the activity took place.
Use the `project`'s value if you want to have multiple projects `POST` to the same URL.

> Not to be mistaken with the [`Project` type action](#type).

Author: As a non-native speaker I found the first one clearer, but either way works.

### `type`

The `type` value is a combination of the scope and the action itself for that event.
Its scope can be in one of the following categories:

> - [Project](#project-event-value)
> - [Environment](#environment-event-value)
> - [Integration](#integration-event-value)

Reviewer: :hammer_and_wrench: 
```suggestion
- [Project](#project-events)
- [Environment](#environment-events)
- [Integration](#integration-events)
```

#### `Project` event value

This value describes the action that happened on a given project, its value can be any from the following table.

| Name | Description |
|------|-------------|
| `project.modify.title` | The human-friendly title of the project has been changed. |

> | `project.create` | A project has been created. Although it appears in the activity feed exactly once, it's not sent via a webhook as it always happens before a webhook can be configured. |

Author: That's correct, that was unneeded information. Removed it!

| `project.variable.create` | A new project variable has been created. |
| `project.variable.delete` | A project variable has been deleted. |
| `project.variable.update` | A project variable has been modified. |

For an overview of the different actions that can take place, see [the type](#type).

#### `Environment` event value

This value describes the action that happened on a project's environment, its value can be any from the following table.

| Name | Description |
|------|-------------|
| `environment.domain.create` | A new domain has been associated with the environment. |
| `environment.domain.delete` | A domain associated with the environment has been removed. |
| `environment.domain.update` | A domain associated with the environment has been updated, such as having its SSL certificate modified. |
| `environment.backup` |  A user triggered a [backup](../../environments/backup.md). |
| `environment.backup.delete` | A user deleted a [backup](../../environments/backup.md) |
| `environment.restore` | A user restored a [backup](../../environments/backup.md). |
| `environment.push` | A user has pushed code to a branch, either existing or new. |

> | `environment.branch` | A new branch has been created via the Console. (A branch created via a push shows up as `environment.push`.) |

Reviewer: It is the same.
```suggestion
| `environment.branch` | A new branch has been created via the CLI, Console, or API. A branch created via a push shows up as `environment.push`. |
```

> | `environment.activate` | A branch has been "activated", and an environment created for it. |

Reviewer: :hammer_and_wrench: 
```suggestion
| `environment.activate` | The environment has been made [active](../../other/glossary.md#active-environment). |
```

| `environment.initialize` | The default branch of the project has just been initialized with its first commit. |
| `environment.deactivate` | A branch has been "deactivated". The code is still there, but the environment was destroyed. |
| `environment.synchronize` | An environment has had its data and/or code re-copied from its parent environment. |
| `environment.merge` | A branch was merged through the Console or Platform.sh API. A basic Git merge doesn't trigger this event. |
| `environment.redeploy` | An environment was redeployed. |

> | `environment.delete` | A branch was deleted. |

Reviewer: Two things:
1. It's slightly confusing for an `environment.delete` branch to be describing a `branch`.
2. There are inconsistencies with how `delete` is used in the Console and CLI. Sometimes it deletes an environment and sometimes it just deletes data. It can be confusing, so I'm not sure what's covered by this event.

> | `environment.delete` | A branch was deleted. |

Author: This deletes the environment's branch code through Git.

| `integration.bitbucket_server.fetch` | Changes in BitBucket repository have been pulled. |
| `integration.bitbucket_server.register_hooks` | Integration hook have been registered on BitBucket. |
| `integration.github.fetch` | Changes in GitHub repository have been pulled. |
| `integration.gitlab.fetch` | Changes in GitLab repository have been pulled. |
| `integration.health.email` | Health event sent by email. |
| `integration.health.pagerduty` | Health event sent to PagerDuty. |
| `integration.health.slack` | Health event sent to slack. |
| `integration.webhook` | Webhook triggered. |
| `integration.script` | An activity script has run. |

For an overview of the different actions that can take place, see [the type](#type).

### `environments`

An array listing the `environments` that were involved in the activity.
This is usually a single-value element.

### `state`

The current state of the activity.
Its value can be `pending`, `in_progress`, or `complete`.

### `completion_percent`

The completion percentage value of the activity.

### `result`

Whether the activity itself completed successfully or not.
Its value is `success` if all went as planned.

> The result doesn't take into account if your cron task executed properly or if the (re)deploy ran successfully.

Reviewer: I'm just not certain where the line between the activity and anything else is. What's the activity in that case? Would it just be because someone didn't include `-e` in the command or something else? I know deploy commands can fail and still be marked as success. But not sure that means "all went as planned"

| `payload.commits` | The git metadata for the changes. |
| `payload.commits_count` | The amount of git commits.  |
| `payload.deployment` | See [`deployment`](#deployment). |

#### `user`

This element contains information about the Platform.sh user that triggered the activity.

| Name | Description |
|------|-------------|
| `payload.user.created_at` | The date the user was created. |
| `payload.user.display_name` | The human-friendly name of the user that triggered the activity. |
| `payload.user.id` | The user ID of the user. |
| `payload.user.updated_at` | The date that user was updated. |

#### `environment`

This element contains information about the environment itself,
its settings and cron commands that have been run,
after the action has taken place.
The most notable values of this element are:

| Name | Description |
|------|-------------|
| `payload.environment.name` | The name of the branch. |
| `payload.environment.machine_name` | The name of the environment. |
| `payload.environment.type` | The type of the environment. |
| `payload.environment.head_commit` | The Git commit ID that triggered the event. |

> | `payload.environment.edge_hostname` | The edge hostname of the environment. |

Author: Should hopefully be clearer now, added a note to the glossary.
